### PR TITLE
Fix #1861: do not use backslashes as path-separators in CSS links

### LINKED ISSUE #1861: Path separators in YAML css parameter rendered as backslashes in HTML hrefs under Windows
Path separators in the `css` field in YAML head matter are converted to backslashes when rendering rmarkdown to HTML formats under Windows. This leads to HTML files with backslashes in the hrefs. This can be fixed with a few minor code changes.

### Expected behavior

When RMarkdown documents are rendered as HTML, `href` fields use forward slashes as path separators

### Observed behavior 

Under windows, `href` fields for links generated from the `css` parameter use backslashes as path separators.

### Related issues and PRs

This is similar to #512, #1289, #1293, and #1302.

### Additional comments

I will submit a PR (changing 3 lines of code) as a suggested fix. See <https://github.com/jonathan-g/rmarkdown/tree/css-slash-fix>

I am not submitting a reprex because it would involve multiple rmarkdown source files (`.Rmd` and `.css`) to reproduce this behavior. Here is an example of a snippet from rmarkdown YAML head matter that will reproduce this:
``` yaml
css: "css/style.css"
output:
  html_document:
    self_contained: false
```
This results in a line in the generated HTML file
```
<link rel="stylesheet" href="css\style.css" type="text/css" />
```

### Session Info
``` r
R version 4.0.2 (2020-06-22)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363), RStudio 1.3.1054

Locale:
  LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
  LC_NUMERIC=C                           LC_TIME=English_United States.1252    

Package version:
  base64enc_0.1.3 digest_0.6.25   evaluate_0.14   glue_1.4.1      graphics_4.0.2  grDevices_4.0.2 highr_0.8       htmltools_0.5.0 jsonlite_1.7.0 
  knitr_1.29      magrittr_1.5    markdown_1.1    methods_4.0.2   mime_0.9        rlang_0.4.7     rmarkdown_2.3.2 stats_4.0.2     stringi_1.4.6  
  stringr_1.4.0   tinytex_0.24    tools_4.0.2     utils_4.0.2     xfun_0.15       yaml_2.2.1     

Pandoc version: 2.10
```

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('rmarkdown')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('rstudio/rmarkdown')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

### PR DESCRIPTION
**Pull request by jonathan-g**
This is a proposed fix for #1861. It is also related to PR #1302.

Fix path separators for `css` parameters in YAML front matter for HTML files under Windows. Previously, these were converted to backslashes even if they appeared in the `css` argument as forward slashes. Now `hrefs` for `css` links are generated with forward slashes under Windows.

I have emailed a signed contributor agreement.

### CODE REVIEW

File: R/html_document.R [hunk 1]
```diff
@@ -295,7 +295,7 @@ html_document <- function(toc = FALSE,
 
   # additional css
   for (css_file in css)
-    args <- c(args, "--css", pandoc_path_arg(css_file))
+    args <- c(args, "--css", pandoc_path_arg(css_file, backslash=FALSE))
```
Comment (yihui) [inline R/html_document.R:298]:
```suggestion
    args <- c(args, "--css", pandoc_path_arg(css_file, backslash = FALSE))
```

File: R/ioslides_presentation.R [hunk 2]
```diff
@@ -279,7 +279,7 @@ ioslides_presentation <- function(number_sections = FALSE,
 
   # additional css
   for (css_file in css)
-    args <- c(args, "--css", pandoc_path_arg(css_file))
+    args <- c(args, "--css", pandoc_path_arg(css_file, backslash=FALSE))
```
Comment (yihui) [inline R/ioslides_presentation.R:282]:
```suggestion
    args <- c(args, "--css", pandoc_path_arg(css_file, backslash = FALSE))
```

File: R/slidy_presentation.R [hunk 3]
```diff
@@ -103,7 +103,7 @@ slidy_presentation <- function(number_sections = FALSE,
 
   # additional css
   for (css_file in css)
-    args <- c(args, "--css", pandoc_path_arg(css_file))
+    args <- c(args, "--css", pandoc_path_arg(css_file, backslash=FALSE))
```
Comment (yihui) [inline R/slidy_presentation.R:106]:
```suggestion
    args <- c(args, "--css", pandoc_path_arg(css_file, backslash = FALSE))
```

File: NEWS.md [hunk 4]
```diff
@@ -6,6 +6,8 @@ rmarkdown 2.4
     - New argument `lua_filters` in `pandoc_options()` to pass the Lua filter paths to use with a format. This allow output format authors to add filters for a custom format using the `pandoc` argument of `output_format()` and to get filters from a format using `fmt$pandoc$lua_filters`.
     - The Lua filters of an output format are now passed to Pandoc in `render()`. By default, they are passed to Pandoc before any other format-defined or user-defined Pandoc arguments (usually via the `pandoc_args` option of an output format). This ensures that filters of an output format are executed first. To change the default, you need to deal with it in a custom format (i.e., modify the elements in `fmt$pandoc$lua_filters`, such as reordering them).
 
+- Fix path separators for "css" parameters in YAML frontmatter for HTML files under Windows. Previously, these were converted to backslashes even if they appeared in the "css" (thanks, @jonathan-g, #1862)
```
Comment (yihui) [inline NEWS.md:9]:
```suggestion
- Fixed the path separators for the `css` parameter in YAML frontmatter for HTML output files under Windows. Previously, forward slashes in `css` paths were converted to backslashes (thanks, @jonathan-g, #1862).
```

**Review by yihui — APPROVED** [review]
It's unbelievable that no one has discovered this problem before... Anyway, thanks for the fix!

**Comment by cderv** [comment]
> It's unbelievable that no one has discovered this problem before... Anyway, thanks for the fix!

I think this is because browser are clever and it still works with backslashes. I am on windows and never got an issue when included custom css

Thanks for the fix!

**Comment by jonathan-g** [comment]
Yeah. If you stay within any given OS< this is generally not a problem. I ran into it in a situation where I built a web page on Windows and then pushed it to a Linux system, where the backslashes became a problem.

**Comment by cderv** [comment]
That is fortunate for us then ! This was well hidden ! 
Thank you again !

**Pull request merged**